### PR TITLE
[4.0] GSoC 2018 Enhance Users Project

### PR DESCRIPTION
**Pull request by alexandraciobica**
### Summary of Changes
All the changes regard the users component.

Site:
 - Added two new views:
   - A view named `users` with a list of users in a user group taking into consideration access levels when displaying 
   - A view named `user` that displays the details of a chosen user, including a contact form similar to the one from contact component.
- Changed the router so that it displays the links for these pages correctly

Admin:
- Added `Access` field to the users list - the value can be changed from `Users: Edit` > `Account Details` tab or using batch processing
- Added Contact form to the user - the options can be changed from `Users: Edit` or global options > `Form` tab
- Created two new menu item types: 
   - one that displays a list of users based on groups
   - one that displays the details view of the selected user

### Testing Instructions

I wrote this JDocs article and I think it explains pretty well how to test everything. https://docs.joomla.org/J4.x:Users_List_and_Details_Views

Mentors: @bembelimen and @laoneo

### CODE REVIEW

**Comment by brianteeman** [comment]
(updated title to show it is for 4.0)

File: administrator/components/com_users/config.xml [hunk 1]
```diff
@@ -184,6 +184,113 @@
...
+	>
+
+		<field
+			name="captcha"
+			type="plugins"
+			label="COM_USERS_CONFIG_FIELD_CAPTCHA_LABEL"
+			folder="captcha"
+			filter="cmd"
+			useglobal="true"
+		>
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: indent one tab

File: administrator/components/com_users/config.xml [hunk 2]
```diff
@@ -184,6 +184,113 @@
...
+			<option value="0">JOPTION_DO_NOT_USE</option>
+		</field>
+
+		<field
+			name="show_email_form"
+			type="radio"
+			class="switcher"
+			label="COM_USERS_FIELD_EMAIL_SHOW_FORM_LABEL"
+			default="1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: indent one tab

File: administrator/components/com_users/config.xml [hunk 3]
```diff
@@ -184,6 +184,113 @@
...
+		</field>
+
+		<field
+			name="show_email_copy"
+			type="radio"
+			label="COM_USERS_FIELD_EMAIL_EMAIL_COPY_LABEL"
+			default="0"
+			class="switcher"
+			showon="show_email_form:1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: indent one tab

File: administrator/components/com_users/config.xml [hunk 4]
```diff
@@ -184,6 +184,113 @@
...
+			rows="3"
+			cols="30"
+			showon="show_email_form:1"
+		/>
+
+		<field
+			name="validate_session"
+			type="radio"
+			class="switcher"
+			label="COM_USERS_FIELD_CONFIG_SESSION_CHECK_LABEL"
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: move label above class

File: administrator/components/com_users/config.xml [hunk 5]
```diff
@@ -184,6 +184,113 @@
...
+			default="1"
+			showon="show_email_form:1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: indent one tab

File: administrator/components/com_users/config.xml [hunk 6]
```diff
@@ -184,6 +184,113 @@
...
+			<option value="0">JNO</option>
+			<option value="1">JYES</option>
+		</field>
+
+		<field
+			name="custom_reply"
+			type="radio"
+			class="switcher"
+			label="COM_USERS_CONFIG_FIELD_CUSTOM_REPLY_LABEL"
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: move label above class

File: administrator/components/com_users/config.xml [hunk 7]
```diff
@@ -184,6 +184,113 @@
...
+			default="0"
+			showon="show_email_form:1"
+		>
```
Comment (brianteeman) [inline administrator/components/com_users/config.xml]: indent one tab

File: administrator/components/com_users/forms/user.xml [hunk 8]
```diff
@@ -203,5 +210,83 @@
 
 		</fieldset>
 
+		<fieldset name="email" label="COM_USERS_FIELDSET_CONTACT_LABEL">
+
+			<field
+					name="show_email_form"
```
Comment (brianteeman) [inline administrator/components/com_users/forms/user.xml]: you need to check the tabs on all of this file and consider the previous comments about codestyle

File: administrator/components/com_users/tmpl/users/default.php [hunk 9]
```diff
@@ -148,6 +151,9 @@
 										<?php echo nl2br($item->group_names); ?>
 									<?php endif; ?>
 								</td>
+                                <td class="small d-none d-md-table-cell text-center">
```
Comment (brianteeman) [inline administrator/components/com_users/tmpl/users/default.php]: please indent to match the surrounding code

File: administrator/language/en-GB/en-GB.com_users.ini [hunk 10]
```diff
@@ -15,13 +15,22 @@ COM_USERS_BATCH_SET="Set To Group"
...
 COM_USERS_CONFIG_FIELD_CAPTCHA_LABEL="Captcha"
 COM_USERS_CONFIG_FIELD_CHANGEUSERNAME_LABEL="Change Username"
+COM_USERS_CONFIG_FIELD_CONTACT_FORM="Form"
+COM_USERS_CONFIG_FIELD_CUSTOM_REPLY_LABEL="Custom Reply"
 COM_USERS_CONFIG_FIELD_FRONTEND_LANG_LABEL="Frontend Language"
 COM_USERS_CONFIG_FIELD_FRONTEND_RESET_COUNT_LABEL="Maximum Reset Count"
 COM_USERS_CONFIG_FIELD_FRONTEND_RESET_TIME_LABEL="Reset Time"
 COM_USERS_CONFIG_FIELD_FRONTEND_USERPARAMS_LABEL="Frontend User Parameters"
 COM_USERS_CONFIG_FIELD_GUEST_USER_GROUP_LABEL="Guest User Group"
+COM_USERS_CONFIG_FIELD_INDIVIDUAL_CONTACT_DESC="These settings apply for single User unless they are changed for a specific User."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_users.ini]: That doesn't make sense - /me confused

File: administrator/language/en-GB/en-GB.com_users.sys.ini [hunk 11]
```diff
@@ -24,8 +28,12 @@ COM_USERS_NOTE_VIEW_EDIT_TITLE="Create User Note"
 COM_USERS_TAGS_CATEGORY="User Note Category"
 COM_USERS_USERS_VIEW_DEFAULT_DESC="Shows a List of Users"
 COM_USERS_USERS_VIEW_DEFAULT_TITLE="Users"
+COM_USERS_USER_DETAILS_VIEW_DEFAULT_DESC="Displays the details of the chosen user"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_users.sys.ini]: please add a . at the end - all descriptions should be sentences

**Review by brianteeman — CHANGES_REQUESTED** [review]

**Comment by TobsBobs** [comment]
If this is ready for testing, please give a hint.

File: administrator/components/com_users/config.xml [hunk 12]
```diff
@@ -184,6 +184,113 @@
...
+			<option value="0">JNO</option>
+			<option value="1">JYES</option>
+		</field>
+
+		<field
+			name="custom_reply"
+			type="radio"
+			class="switcher"
```
Comment (Quy) [inline administrator/components/com_users/config.xml]: Per [Joomla Coding Standards](https://developer.joomla.org/coding-standards/xml.html):

> The four attributes `name`, `type`, `label` and `description` should be written in this order and at the top of the element definition.

File: components/com_users/View/User/HtmlView.php [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+	 *
+	 * @throws  \Exception
+	 */
+	protected function _prepareDocument()
+	{
+		$app     = Factory::getApplication();
+		$pathway = $app->getPathway();
+
+		$pathway->addItem($this->item->name, '');
+
```
Comment (Quy) [inline components/com_users/View/User/HtmlView.php]: Remove blank line.

File: administrator/components/com_users/forms/user.xml [hunk 14]
```diff
@@ -203,5 +210,83 @@
...
 		</fieldset>
 
+		<fieldset name="email" label="COM_USERS_FIELDSET_CONTACT_LABEL">
+
+			<field
+				name="show_email_form"
+				type="list"
+				label="COM_USERS_FIELD_EMAIL_SHOW_FORM_LABEL"
+				useglobal="true"
+
```
Comment (Quy) [inline administrator/components/com_users/forms/user.xml]: Remove blank line.

File: components/com_users/Controller/UserController.php [hunk 15]
```diff
@@ -362,4 +381,195 @@ public function resend()
...
+	 *
+	 * @since   1.6.4
+	 */
+	private function _sendEmail($data, $contact, $copy_email_activated)
+	{
+		$app = $this->app;
+
+		if ($contact->email == '' && $contact->id != 0)
+		{
+			$contact_user      = User::getInstance($contact->id);
```
Comment (Quy) [inline components/com_users/Controller/UserController.php]: Align `=`

File: components/com_users/Model/UserModel.php [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+		$pk = $app->input->getInt('id');
+		$this->setState('user.id', $pk);
+
+		$offset = $app->input->getUInt('limitstart');
+		$this->setState('list.offset', $offset);
+
+		// Load the parameters.
+		$params = $app->getParams();
+		$this->setState('params', $params);
+
```
Comment (Quy) [inline components/com_users/Model/UserModel.php]: Remove blank line

File: components/com_users/Model/UserModel.php [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+			}
+
+			$loggedUser = Factory::getUser();
+			$groups = $loggedUser->getAuthorisedViewLevels();
+
+			$registry = new Registry($user->params);
+			$user->params = $this->getState('params');
+			$user->params = clone $this->getState('params');
+			$user->params->merge($registry);
+
```
Comment (Quy) [inline components/com_users/Model/UserModel.php:93]: Remove blank line

File: components/com_users/layouts/field/render.php [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+$class     = $field->params->get('render_class');
+$showLabel = $field->params->get('showlabel');
+
+if ($field->context == 'com_users.contact')
+{
+	// Prepare the value for the contact form mail
+	$value = html_entity_decode($value);
+
+	echo ($showLabel ? $label . ': ' : '') . $value . "\r\n";
+	return;
```
Comment (Quy) [inline components/com_users/layouts/field/render.php]: Add blank line above

File: components/com_users/router.php [hunk 19]
```diff
@@ -40,10 +43,73 @@ public function __construct($app = null, $menu = null)
...
+
+		if ($this->noIDs)
+		{
+			list($id, $segmentList) = explode('-', $segment, 2);
+
+			return (int) $id;
+		}
+
+		return (int) $segment;
+
```
Comment (Quy) [inline components/com_users/router.php]: Remove blank line

File: components/com_users/tmpl/users/default.xml [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+
+	<!-- Add fields to the parameters object for the layout. -->
+	<fields name="params">
+		<fieldset name="basic" label="COM_USERS_GROUP_LABEL">
+			<field
+				name="show_group_title"
+				type="list"
+				label="COM_USERS_GROUP_FIELD_TITLE_LABEL"
+				useglobal="true"
+			>
```
Comment (Quy) [inline components/com_users/tmpl/users/default.xml]: Indent

File: administrator/components/com_users/tmpl/users/default.php (hunk 9 above)
Comment (alexandraciobica) [reply administrator/components/com_users/tmpl/users/default.php]: This is aligned correctly in my editor, I don't know why here is not..

Comment (mbabker) [reply administrator/components/com_users/tmpl/users/default.php]: You're using spaces for indents, the rest of the file is using tabs.

**Comment by brianteeman** [comment]
I am confused by the objectives of this new access column 

- what is the problem that is being solved
- what can I do now I couldn't do before
- what happens with existing users. the column is blank for them
- our access and viewlevels are confusing enough atm and I "think" this makes it worse - for example I can create a new user in the admin and select the access as super users. But that is not the same as changing their group to Super users
chrome_2018-08-08_22-43-14

**Comment by brianteeman** [comment]
### contact form 

Isn't this duplicating the functionality already possible with the user - contact creator plugin?

**Comment by brianteeman** [comment]
New menu items

I couldnt test this as the two options were not present in the list of available menu types that can be created

**Comment by laoneo** [comment]
The goal of this project was to provide front end menu list items for users and a detail page. This solves tow issues:

- The long outstanding problem that the user menu item is not working on the front end
- That we can decouple com_contact as you need com_contact to show a list of users on the front end.

The access flag is a security measure that the admin can define who can see the user on the front end. I was the driver behind that flag and not to have just a yes no button if the user details should be available on the front or not.

**Comment by SharkyKZ** [comment]
>The access flag is a security measure that the admin can define who can see the user on the front end.

Per-user setting should be set by the user in user profile. Admin could set the global setting in `com_users` to allow/disallow frontend profile viewing. Need to think about this from user perspective too.

**Comment by brianteeman** [comment]
> The goal of this project was to provide front end menu list items for users and a detail page. 

Why? There are many extensions for that. Is it really something that the majority of websites would want? If not then it doesnt belong in the core

> That we can decouple com_contact as you need com_contact to show a list of users on the front end.

If you need a list at all - see above.. And as I stated this is duplicating functionality available with the contact creator plugin which allows much more advanced use. 

> The access flag is a security measure that the admin can define who can see the user on the front end. I was the driver behind that flag and not to have just a yes no button if the user details should be available on the front or not.

That is completely unclear from the UI. I read it as what is the access of this user. Again completely unnecessary to add to all users when the contacts can be used for this

Sorry but this makes no sense to me at all for a core contribution. It's not about the code but about should it be in the core

**Comment by laoneo** [comment]
> Is it really something that the majority of websites would want?

Nobody can answer that, not even you.

> And as I stated this is duplicating functionality available with the contact creator plugin

The creator plugin creates a contact when a user is created and has nothing to do with the contact form. Means we have even more bloat, as you have then a contact per user.
I was talking with several people before GsoC started about that project and the contact form is a feature some do need. So the student duplicated that functionality. But this is only till the decoupling is finished.

As I said the overall goal is to be able to decouple com_contact. With custom fields in combination with the user component, this is possible now. Most maintainers do agree that we should slim down the core so this is a step into that direction. If you want to have a contact extension then take one from JED and not the opposite.

**Comment by brianteeman** [comment]
>> Is it really something that the majority of websites would want?

> Nobody can answer that, not even you.

So go to the joomla showcase site and show me how many sites have a user list and then compare that to how many have a contact list

This is not slimming down the core it is bloating it. Sorry not wasting my words on this any more. It is a useless feature that doesnt belong in core.

**Comment by mbabker** [comment]
Tend to agree here.  Having a core frontend user profile view IMO is a nice to have capability, but I don't think it needs to come with all the extra stuff being proposed here, especially the bits related to access that just adds even more confusing things to an already confusing management interface.

**Comment by rdeutz** [comment]
I am sorry but I also don't see this as a needed addition to the core package

**Comment by laoneo** [comment]
PD (PLT) accepted these projects so I don't see a reason why we are discussing again if this should get into core or not. What we can discus here if the code is correct and all the features are implemented the right way.

@mbabker we went with access levels basically because of security reasons. Like that an admin or the user has then to actively activate a user for front end listing.

**Comment by brianteeman** [comment]
Sorry but just because a project was chosen shouldn't mean we have to merge the resultant code. That's never been the case before. Surely as a developer you understand that what might appear to be a good idea on paper isn't necessarily a good idea on practice.

**Comment by ghost** [comment]
@brianteeman that Decision should be made before chose Project.

**Comment by brianteeman** [comment]
Gsoc is about introducing students to open source.

**Comment by laoneo** [comment]
This is only part of GsoC. It should also help the open source projects. If we would do something which will not get accepted in core, then there is no reason to ever start that project. It is a big amount of time the student and the mentors do put into that project. And when it gets trashed at the end then it leads to big frustrations. The GsoC admins learned from the past and did put the proposals into Production for voting, that we can add something valuable to the core and not just waste time and energy.

**Comment by ggppdk** [comment]
This 
unlike other already merge features

1. does not break anything (or at least any breaks / bugs should be easy to fix)
2. does little to complicate common website tasks / UI
3. does little to hurt or compete with 3rd party (i mean 'social' extensions) even if one would create a layout overrides to add more stuff to it, still it will not compete with them ... much
4. will be easy to maintain

Not accepting also considering the above points, is a bad message as @laneo said

I understand the only argument against this (which i admit is a good argument) is that it will be uncommonly useful 
if so the project should not have started at all

I mean the reason for rejecting this was "known" before the project started,
unlike reasons 1, 2, 4 that may have risen after the project completed

**Comment by brianteeman** [comment]
> A view named users with a list of users in a user group taking into consideration access levels when displaying

I do not see any menu type for this so cannot test

> A view named user that displays the details of a chosen user, including a contact form similar to the one from contact component.

The form is not available until you go into com-users options and save 

The view will not display until you go into com_users user and save every user so that the access field is created. Otherwise you have access denied


The form has no language strings 
chrome_2018-08-17_10-46-57

**Comment by ghost** [comment]
Applying PR got:

screen shot 2018-08-17 at 12 06 56

**Comment by mbabker** [comment]
> he GsoC admins learned from the past and did put the proposals into Production for voting, that we can add something valuable to the core and not just waste time and energy.

Just because the project proposal was accepted (which at the time was nothing more than a couple paragraph abstract at best) does NOT mean that the resultant code MUST be accepted.  The end result should be evaluated the same as every pull request and not forced through simply because it has the GSoC label on it.

**Comment by brianteeman** [comment]
And if you could please evaluate the comments about the code and make it work so it can be tested then it would obviously help your case to get this accepted. Right now as I have written twice before there is NO code to display the user list

**Comment by brianteeman** [comment]
Sorry but I am closing this. There has been no response to requests to at least get it into a testable state

**Comment by laoneo** [comment]
The student promised to have a look. Perhaps she needs some more time.

**Comment by brianteeman** [comment]
It can always be re-opened if that ever happens

**Pull request closed without merging**